curvefs/src/metaserver/metaserver.cpp

```diff
@@ -186,21 +185,54 @@ void Metaserver::Run() {
     LOG_IF(FATAL, server_->AddService(metaService_.get(),
                                       brpc::SERVER_DOESNT_OWN_SERVICE) != 0)
         << "add metaserverService error";
-
     LOG_IF(FATAL, server_->AddService(copysetService_.get(),
                                       brpc::SERVER_DOESNT_OWN_SERVICE) != 0)
         << "add copysetservice error";
 
+    butil::ip_t ip;
+    LOG_IF(FATAL, 0 != butil::str2ip(options_.ip.c_str(), &ip))
+        << "convert " << options_.ip << " to ip failed";
+    butil::EndPoint listenAddr(ip, options_.port);
+
     // add raft-related service
     copysetNodeManager_->AddService(server_.get(), listenAddr);
 
-    // start rpc server
+    // start internal rpc server
     brpc::ServerOptions option;
     if (options_.bthreadWorkerCount != -1) {
         option.num_threads = options_.bthreadWorkerCount;
     }
     LOG_IF(FATAL, server_->Start(listenAddr, &option) != 0)
-        << "start brpc server error";
+        << "start internal brpc server error";
+
+    // add external server
+    externalServer_ = absl::make_unique<brpc::Server>();
+    if (options_.enableExternalServer) {
+        LOG_IF(FATAL, externalServer_->AddService(metaService_.get(),
+            brpc::SERVER_DOESNT_OWN_SERVICE) != 0)
+            << "add metaserverService error";
+        LOG_IF(FATAL, externalServer_->AddService(copysetService_.get(),
+            brpc::SERVER_DOESNT_OWN_SERVICE) != 0)
+            << "add copysetservice error";
+
+        butil::ip_t ip;
+        LOG_IF(FATAL, 0 != butil::str2ip(
+            registerOptions_.metaserverExternalIp.c_str(), &ip))
+            << "convert " << registerOptions_.metaserverExternalIp
+            << " to ip failed";
+        butil::EndPoint listenAddr(ip, registerOptions_.metaserverExternalPort);
+
+        // add raft-related service
+        copysetNodeManager_->AddService(externalServer_.get(), listenAddr);
```

Comment:
raft-related services should only be added to the internal server.

previous ChunkServer implementation adds raft-related service because it has compatible problem, there're lots of old clients that can upgrade.

and in `CopysetNodeManager::AddService`, https://github.com/opencurve/curve/blob/895fa84778dc872174f42179f2d2b5d565152fa2/curvefs/src/metaserver/copyset/copyset_node_manager.cpp#L221
you should add `RaftCliService2` into external server.

Reply:
> raft-related services should only be added to the internal server.
> 
> previous ChunkServer implementation adds raft-related service because it has compatible problem, there're lots of old clients that can upgrade.
> 
> and in `CopysetNodeManager::AddService`,
> 
> https://github.com/opencurve/curve/blob/895fa84778dc872174f42179f2d2b5d565152fa2/curvefs/src/metaserver/copyset/copyset_node_manager.cpp#L221
> 
> 
> you should add `RaftCliService2` into external server.

fixed